File: backend/ITSAssetTracker.API/Services/CsvProcessingService.cs
using System.Globalization;
using CsvHelper;
using CsvHelper.Configuration;
using ITSAssetTracker.API.Data;
using ITSAssetTracker.API.Models;
using ITSAssetTracker.API.Models.DTOs;
using Microsoft.EntityFrameworkCore;

namespace ITSAssetTracker.API.Services;

public class CsvProcessingService : ICsvProcessingService
{
    private readonly AssetTrackerDbContext _context;
    private readonly ILogger<CsvProcessingService> _logger;
    private readonly IEmailService _emailService;
    private readonly IConfiguration _configuration;

    public CsvProcessingService(
        AssetTrackerDbContext context,
        ILogger<CsvProcessingService> logger,
        IEmailService emailService,
        IConfiguration configuration)
    {
        _context = context;
        _logger = logger;
        _emailService = emailService;
        _configuration = configuration;
    }

    public async Task<CsvUploadResultDto> ProcessCsvUpload(Stream csvStream, string uploadedBy)
    {
        var result = new CsvUploadResultDto();
        var lowStockItems = new List<Inventory>();

        try
        {
            using var reader = new StreamReader(csvStream);
            var config = new CsvConfiguration(CultureInfo.InvariantCulture)
            {
                HeaderValidated = null,
                MissingFieldFound = null
            };

            using var csv = new CsvReader(reader, config);
            var records = csv.GetRecords<InventoryCsvRow>().ToList();

            _logger.LogInformation("Processing {Count} CSV records uploaded by {User}", records.Count, uploadedBy);

            foreach (var record in records)
            {
                try
                {
                    await ProcessCsvRow(record, uploadedBy, lowStockItems);
                    result.SuccessCount++;
                }
                catch (Exception ex)
                {
                    result.FailureCount++;
                    result.Errors.Add($"Row with ItemNumber '{record.ItemNumber}': {ex.Message}");
                    _logger.LogError(ex, "Error processing CSV row for item {ItemNumber}", record.ItemNumber);
                }
            }

            await _context.SaveChangesAsync();

            // Send email alerts for low stock items
            if (lowStockItems.Any())
            {
                var appBaseUrl = _configuration["AppSettings:BaseUrl"] ?? "http://localhost:3000";
                await _emailService.SendLowStockAlerts(lowStockItems, appBaseUrl);

                result.LowStockAlerts = lowStockItems
                    .Select(i => $"{i.ItemNumber} - {i.HardwareDescription} (Qty: {i.CurrentQuantity}, Threshold: {i.MinimumThreshold})")
                    .ToList();
            }

            _logger.LogInformation("CSV upload completed. Success: {Success}, Failures: {Failures}, Low Stock Alerts: {Alerts}",
                result.SuccessCount, result.FailureCount, lowStockItems.Count);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing CSV upload");
            throw;
        }

        return result;
    }

    private async Task ProcessCsvRow(InventoryCsvRow row, string uploadedBy, List<Inventory> lowStockItems)
    {
        var existingItem = await _context.Inventories
            .FirstOrDefaultAsync(i => i.ItemNumber == row.ItemNumber);

        if (existingItem != null)
        {
            // Update existing item
            var previousQuantity = existingItem.CurrentQuantity;

            existingItem.HardwareDescription = row.HardwareDescription;
            existingItem.HardwareType = row.HardwareType;
            existingItem.Cost = row.Cost;
            existingItem.MinimumThreshold = row.MinimumThreshold;
            existingItem.ReorderAmount = row.ReorderAmount;
            existingItem.CurrentQuantity = row.CurrentQuantity;
            existingItem.LastModifiedBy = row.AuditUser ?? uploadedBy;
            existingItem.LastModifiedDate = DateTime.UtcNow;

            // Create audit entry if quantity changed
            if (previousQuantity != row.CurrentQuantity)
            {
                var auditEntry = new AuditHistory
                {
                    ItemId = existingItem.Id,
                    PreviousQuantity = previousQuantity,
                    NewQuantity = row.CurrentQuantity,
                    ChangedBy = row.AuditUser ?? uploadedBy,
                    ChangeDate = DateTime.UtcNow,
                    ServiceNowTicketUrl = row.ServiceNowTicketUrl
                };

                _context.AuditHistories.Add(auditEntry);
            }

            // Check if item is now below threshold
            if (existingItem.CurrentQuantity <= existingItem.MinimumThreshold)
            {
                lowStockItems.Add(existingItem);
            }
        }
        else
        {
            // Create new item
            var newItem = new Inventory
            {
                ItemNumber = row.ItemNumber,
                HardwareDescription = row.HardwareDescription,
                HardwareType = row.HardwareType,
                Cost = row.Cost,
                MinimumThreshold = row.MinimumThreshold,
                ReorderAmount = row.ReorderAmount,
                CurrentQuantity = row.CurrentQuantity,
                LastModifiedBy = row.AuditUser ?? uploadedBy,
                LastModifiedDate = DateTime.UtcNow
            };

            _context.Inventories.Add(newItem);
            await _context.SaveChangesAsync(); // Save to get the ID

            // Create audit entry for new item
            var auditEntry = new AuditHistory
            {
                ItemId = newItem.Id,
                PreviousQuantity = 0,
                NewQuantity = row.CurrentQuantity,
                ChangedBy = row.AuditUser ?? uploadedBy,
                ChangeDate = DateTime.UtcNow,
                ServiceNowTicketUrl = row.ServiceNowTicketUrl
            };

            _context.AuditHistories.Add(auditEntry);

            // Check if new item is below threshold
            if (newItem.CurrentQuantity <= newItem.MinimumThreshold)
            {
                lowStockItems.Add(newItem);
            }
        }
    }
}

File: backend-express/src/services/csvProcessing.service.ts
import { parse } from 'csv-parse/sync';
import { Inventory } from '@prisma/client';
import prisma from '../config/prisma';
import { emailService } from './email.service';
import { config } from '../config';
import { sanitizeCsvField, parseNumericField, parseIntegerField } from '../utils/csv-sanitizer';

export interface CsvUploadResult {
  successCount: number;
  failureCount: number;
  errors: string[];
  lowStockAlerts: Array<{
    itemNumber: string;
    description: string;
    currentQuantity: number;
    minimumThreshold: number;
  }>;
}

interface InventoryCsvRow {
  ItemNumber: string;
  AssetType?: string;
  Description: string;
  Category?: string;
  Cost: string;
  MinimumThreshold: string;
  ReorderAmount: string;
  CurrentQuantity?: string;
  HardwareDescription?: string;
  HardwareType?: string;
}

export interface ICsvProcessingService {
  processCsvUpload(csvBuffer: Buffer, uploadedBy: string): Promise<CsvUploadResult>;
  generateCsvTemplate(): string;
}

class CsvProcessingService implements ICsvProcessingService {
  /**
   * Process uploaded CSV file and update inventory
   */
  async processCsvUpload(csvBuffer: Buffer, uploadedBy: string): Promise<CsvUploadResult> {
    const result: CsvUploadResult = {
      successCount: 0,
      failureCount: 0,
      errors: [],
      lowStockAlerts: [],
    };

    try {
      // Parse CSV
      const records = parse(csvBuffer, {
        columns: true,
        skip_empty_lines: true,
        trim: true,
      }) as InventoryCsvRow[];

      if (records.length === 0) {
        result.errors.push('CSV file is empty');
        return result;
      }

      // Process each row
      for (let i = 0; i < records.length; i++) {
        const row = records[i];
        const rowNumber = i + 2; // +2 because of header row and 0-based index

        try {
          // Validate required fields
          if (!row.ItemNumber || !row.Description) {
            result.errors.push(`Row ${rowNumber}: ItemNumber and Description are required`);
            result.failureCount++;
            continue;
          }

          // SECURITY: Sanitize fields to prevent CSV injection
          const itemNumber = sanitizeCsvField(row.ItemNumber);
          const rawDescription = row.Description || row.HardwareDescription || '';
          const description = sanitizeCsvField(rawDescription);
          const assetType = sanitizeCsvField(row.AssetType || 'Hardware');
          const category = row.Category || row.HardwareType
            ? sanitizeCsvField(row.Category || row.HardwareType)
            : null;

          // SECURITY: Validate numeric fields with bounds checking
          const cost = parseNumericField(row.Cost, 'Cost', 0, 999999.99);
          const minimumThreshold = parseIntegerField(row.MinimumThreshold, 'MinimumThreshold', 0, 999999);
          const reorderAmount = parseIntegerField(row.ReorderAmount, 'ReorderAmount', 0, 999999);
          const currentQuantity = row.CurrentQuantity
            ? parseIntegerField(row.CurrentQuantity, 'CurrentQuantity', 0, 999999)
            : 0;

          // Check if item exists (use sanitized itemNumber)
          const existingItem = await prisma.inventory.findUnique({
            where: { itemNumber },
          });

          if (existingItem) {
            // Update existing item
            const previousQuantity = existingItem.currentQuantity;

            await prisma.inventory.update({
              where: { id: existingItem.id },
              data: {
                assetType,
                description,
                category,
                cost,
                minimumThreshold,
                reorderAmount,
                currentQuantity: currentQuantity > 0 ? currentQuantity : existingItem.currentQuantity,
                lastModifiedBy: uploadedBy,
                lastModifiedDate: new Date(),
                // Update legacy fields for backward compatibility
                hardwareDescription: row.HardwareDescription || description,
                hardwareType: row.HardwareType || category,
              },
            });

            // Create audit entry if quantity changed
            if (currentQuantity > 0 && currentQuantity !== previousQuantity) {
              await prisma.auditHistory.create({
                data: {
                  itemId: existingItem.id,
                  previousQuantity,
                  newQuantity: currentQuantity,
                  changedBy: uploadedBy,
                  changeDate: new Date(),
                },
              });
            }

            // Check if item is low stock
            const updatedQuantity = currentQuantity > 0 ? currentQuantity : existingItem.currentQuantity;
            if (updatedQuantity < minimumThreshold) {
              result.lowStockAlerts.push({
                itemNumber,
                description,
                currentQuantity: updatedQuantity,
                minimumThreshold,
              });
            }
          } else {
            // Create new item
            const newItem = await prisma.inventory.create({
              data: {
                itemNumber,
                assetType,
                description,
                category,
                cost,
                minimumThreshold,
                reorderAmount,
                currentQuantity: currentQuantity || 0,
                lastModifiedBy: uploadedBy,
                lastModifiedDate: new Date(),
                hardwareDescription: row.HardwareDescription || description,
                hardwareType: row.HardwareType || category,
              },
            });

            // Create initial audit entry
            if (currentQuantity > 0) {
              await prisma.auditHistory.create({
                data: {
                  itemId: newItem.id,
                  previousQuantity: 0,
                  newQuantity: currentQuantity,
                  changedBy: uploadedBy,
                  changeDate: new Date(),
                },
              });
            }

            // Check if new item is low stock
            if (currentQuantity < minimumThreshold) {
              result.lowStockAlerts.push({
                itemNumber,
                description,
                currentQuantity,
                minimumThreshold,
              });
            }
          }

          result.successCount++;
        } catch (error: any) {
          result.errors.push(`Row ${rowNumber}: ${error.message}`);
          result.failureCount++;
        }
      }

      // Send low stock alerts if any
      if (result.lowStockAlerts.length > 0) {
        try {
          const lowStockItems = await prisma.inventory.findMany({
            where: {
              itemNumber: {
                in: result.lowStockAlerts.map(alert => alert.itemNumber),
              },
            },
          });

          await emailService.sendLowStockAlerts(lowStockItems, config.baseUrl);
        } catch (emailError) {
          console.error('Error sending low stock alerts:', emailError);
          // Don't fail the whole operation if email fails
        }
      }
    } catch (error: any) {
      result.errors.push(`CSV parsing error: ${error.message}`);
      result.failureCount = 1;
    }

    return result;
  }

  /**
   * Generate CSV template for download
   */
  generateCsvTemplate(): string {
    const headers = [
      'ItemNumber',
      'AssetType',
      'Description',
      'Category',
      'Cost',
      'MinimumThreshold',
      'ReorderAmount',
      'CurrentQuantity',
    ];

    const exampleRow = [
      'HW-001',
      'Hardware',
      'Dell Latitude 7420 Laptop',
      'Laptop',
      '1200.00',
      '10',
      '20',
      '15',
    ];

    return `${headers.join(',')}\n${exampleRow.join(',')}\n`;
  }
}

export const csvProcessingService = new CsvProcessingService();
